### PR TITLE
[SPARK-21922] Fix duration always updating when task failed but status is still RUN…

### PR DESCRIPTION
**Pull request by caneGuy**
…NING

## What changes were proposed in this pull request?
When driver quit abnormally which cause executor shutdown and task metrics can not be sent to driver for updating.In this case the status will always be 'RUNNING' and the duration on history UI will be 'CurrentTime - launchTime' which increase infinitely.
We can fix this time by modify time of event log since this time has gotten when `FSHistoryProvider` fetch event log from File System.
And the result picture is uploaded in [SPARK-21922](https://issues.apache.org/jira/browse/SPARK-21922).
How to reproduce?
(1) Submit a job to spark on yarn
(2) Mock an oom(or other case can make driver quit abnormally)  senario for driver
(3) Make sure executor is running task when driver quitting
(4) Open the history server and checkout result
It is not a corner case since there are many such jobs in our current cluster.

## How was this patch tested?
Deploy historyserver and open a job has this problem.

### CODE REVIEW

**Comment by ajbozarth** [comment]
Though this may fix the problem it is 100% the wrong way to do it for multiple reasons:
1. The UI should not touch the FS, FS access it abstracted out so the UI is can work with multiple FS impls
2. This fix assumes it's running the History Server in a class used by both the Web UI and SHS

Without a deeper look at the issue I would instead look at why the problem is occurring and fixing that rather than putting a bandaid on the UI like this. (Note: I've done this before and got the same response, nothing personally :)

I would recommend closing this and opening a new PR once you figure out a new solution.

**Comment by caneGuy** [comment]
Thanks for your recommendation @ajbozarth .Could you put a link for your pr?For the problem you mentioned,i have thought about them.
1、FsHistoryServer will always use FS to get event log
2、For spark ui, my implementation will not access FS.

**Comment by caneGuy** [comment]
@ajbozarth I have updated the implementation which only access FS in FSHistoryServerProvider

File: core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala [hunk 1]
```diff
@@ -142,7 +142,7 @@ private[v1] object AllStagesResource {
       index = uiData.taskInfo.index,
       attempt = uiData.taskInfo.attemptNumber,
       launchTime = new Date(uiData.taskInfo.launchTime),
-      duration = uiData.taskDuration,
+      duration = uiData.taskDuration(),
```
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala]: this is unrelated and unnecessary

**Review by ajbozarth — APPROVED** [review]
Thank you for the update, this is a much better implementation, just one nit otherwise LGTM

File: core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala (hunk 1 above)
Comment (caneGuy) [reply core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala]: Since i changed `taskDuration` below,if do not add `()` a compiler error will occur.

Comment (ajbozarth) [reply core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala]: odd, I thought for sure it'd be fine, then this LGTM

Comment (caneGuy) [reply core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala]: Thanks @ajbozarth

**Comment by caneGuy** [comment]
Ping @kiszk @jerryshao  can you help reivew this?

**Comment by caneGuy** [comment]
@vanzin @zsxwing could you help reivew this?Thanks.

**Comment by jerryshao** [comment]
Is this a problem only in History UI, or it also has issues in Live UI? From my understanding you only pass a last update time for history UI, so is it intended?

Also you mentioned "When executor failed and task metrics have not send to driver,the status will always be 'RUNNING'", is this a bug in scheduler?

**Comment by caneGuy** [comment]
I just appear on history ui since Live ui will use heartbeat metric result @jerryshao(From my understanding) .It appears when Driver exit unnormally,then status of taskinfo can not be updated correctlly.

**Comment by jerryshao** [comment]
So your PR description is quite confusing, would you please elaborate your problem in detail and describe how to reproduce the issue.

**Comment by caneGuy** [comment]
Done @jerryshao

**Comment by jerryshao** [comment]
Still I have a question about history server, is you event log an incomplete event log or completed when you met such issue?

**Comment by caneGuy** [comment]
@jerryshao Thanks for your time. IIUC, event log is completed since driver has not dropped any event of executor which has problem described above.See below,driver only drop two events after shutting down:
`2017-09-11,17:08:57,605 ERROR org.apache.spark.scheduler.LiveListenerBus: SparkListenerBus has already stopped! Dropping event SparkListenerExecutorMetricsUpdate(300,WrappedArray())
2017-09-11,17:08:57,623 INFO org.apache.spark.deploy.yarn.YarnAllocator: Driver requested a total number of 0 executor(s).
2017-09-11,17:08:57,624 INFO org.apache.spark.scheduler.cluster.YarnClusterSchedulerBackend: Shutting down all executors
2017-09-11,17:08:57,624 INFO org.apache.spark.scheduler.cluster.YarnClusterSchedulerBackend: Asking each executor to shut down
2017-09-11,17:08:57,629 ERROR org.apache.spark.scheduler.LiveListenerBus: SparkListenerBus has already stopped! Dropping event SparkListenerExecutorMetricsUpdate(258,WrappedArray())`

And below showing an executor has this problem:
running
running02

Case here is a job running today of our cluster.

**Comment by jerryshao** [comment]
ok to test.

File: core/src/main/scala/org/apache/spark/ui/SparkUI.scala [hunk 2]
```diff
@@ -50,6 +50,7 @@ private[spark] class SparkUI private (
     val operationGraphListener: RDDOperationGraphListener,
     var appName: String,
     val basePath: String,
+    val lastUpdateTime: Long = -1L,
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/ui/SparkUI.scala]: I would like to user `Option[Long] = None` as default value to reflect there's no update time.

**Review by jerryshao — CHANGES_REQUESTED** [review]
Can you please make sure everything is OK in Live UI?

**Comment by jerryshao** [comment]
Looks like I don't have the Jenkins permission to trigger UT 😞 . Let me ping @srowen to trigger the test.

**Comment by caneGuy** [comment]
Yes i have confirmed.Below is a testing job: @jerryshao

File: core/src/main/scala/org/apache/spark/ui/SparkUI.scala (hunk 2 above)
Comment (caneGuy) [reply core/src/main/scala/org/apache/spark/ui/SparkUI.scala]: Update @jerryshao Thanks for your time.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
@jerryshao, for triggering tests on Jenkins, I think this should be added by its admin manually as well if I understood correctly. In my case, I asked this to Josh Rosen before via email privately. I am quite sure you are facing the same issue I (and Holden, Felix and Takuya) met before if I understood correctly.

**Comment by jerryshao** [comment]
Thanks @HyukjinKwon , I will ping Josh about this thing 😄 .

**Comment by caneGuy** [comment]
@jerryshao @HyukjinKwon Jenkins done.Thanks too much.

**Comment by jerryshao** [comment]
Overall LGTM, @ajbozarth can you please review again?

**Comment by ajbozarth** [comment]
I've been following, still LGTM

File: core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala (hunk 1 above)
Comment (jerryshao) [reply core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala]: Here what if we call the REST API on history server to get stage info? Looks like we may still have this issue since we don't have last update time here, what do you think @ajbozarth ?

Comment (caneGuy) [reply core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala]: You are right, @jerryshao .IIUC, the `ui` in `AllStagesResource.scala` is passed from `ApiRootResource` which also create `sparkUI` by `FSHistoryProvider`.So we can also get `lastUpdateTime` from this `ui` in `AllStagesResource` and pass to the `taskDuration` interface.I think it is another problem for REST?Should we fix here?

Comment (jerryshao) [reply core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala]: Yes, if it is not a big change I think it should be fixed here. Because currently with this fix UI and REST API are inconsistent.

File: core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala [hunk 3]
```diff
@@ -47,7 +47,8 @@ private[v1] class AllStagesResource(ui: SparkUI) {
         listener.stageIdToData.get((stageInfo.stageId, stageInfo.attemptId))
       }
     } yield {
-      AllStagesResource.stageUiToStageData(status, stageInfo, stageUiData, includeDetails = false)
+      AllStagesResource.stageUiToStageData(
+        status, stageInfo, stageUiData, includeDetails = false, Some(ui))
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala]: It's not a good idea to pass in `SparkUI` to only get `lastUpdate`, the API looks weird to add this `SparkUI` argument, the fix here only just make it work. It is better to add one more field in `StageUIData` or `TaskUIData` if possible.

File: core/src/main/scala/org/apache/spark/ui/jobs/UIData.scala [hunk 4]
```diff
@@ -97,6 +97,7 @@ private[spark] object UIData {
     var memoryBytesSpilled: Long = _
     var diskBytesSpilled: Long = _
     var isBlacklisted: Int = _
+    var jobLastUpdateTime: Option[Long] = None
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/ui/jobs/UIData.scala]: Is it better to rename to `stageLastUpdateTime` or just `lastUpdateTime`? Since this structure unrelated to job, would be better to not involve "job".

File: core/src/main/scala/org/apache/spark/status/api/v1/OneStageResource.scala [hunk 5]
```diff
@@ -81,7 +83,8 @@ private[v1] class OneStageResource(ui: SparkUI) {
       @DefaultValue("20") @QueryParam("length") length: Int,
       @DefaultValue("ID") @QueryParam("sortBy") sortBy: TaskSorting): Seq[TaskData] = {
     withStageAttempt(stageId, stageAttemptId) { stage =>
-      val tasks = stage.ui.taskData.values.map{AllStagesResource.convertTaskData}.toIndexedSeq
+      val tasks = stage.ui.taskData.values.map{
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/status/api/v1/OneStageResource.scala]: The style should be changed to `map {  AllStagesResource.convertTaskData(_, ui.lastUpdateTime) }`, requires whitespace between `{` and `}`. You could check other similar codes about the style.

File: core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala [hunk 6]
```diff
@@ -69,7 +70,8 @@ private[v1] object AllStagesResource {
       }
 
     val taskData = if (includeDetails) {
-      Some(stageUiData.taskData.map { case (k, v) => k -> convertTaskData(v) } )
+      Some(stageUiData.taskData.map { case (k, v) =>
+        k -> convertTaskData(v, stageUiData.lastUpdateTime) } )
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala]: Sorry I may stated clearly, it should be:

```
Some(stageUiData.taskData.map { case (k, v) =>
  k -> convertTaskData(v, stageUiData.lastUpdateTime) })
```
No whitespace after "}".

Comment (caneGuy) [reply core/src/main/scala/org/apache/spark/status/api/v1/AllStagesResource.scala]: Thanks for your patient.Actually i did not check carefully.

**Comment by jerryshao** [comment]
LGTM, merging to master, if possible to 2.2.

**Comment by jerryshao** [comment]
Cannot cleanly merge to 2.2, so this will only land to master branch.

**Comment by caneGuy** [comment]
@jerryshao Thanks.
